controllers/virtinkmachine_controller.go

```diff
@@ -56,42 +59,90 @@ func (r *VirTinkMachineReconciler) Reconcile(ctx context.Context, req ctrl.Reque
+		return ctrl.Result{}, nil
 	}
 
 	ownerMachine, err := capiutil.GetOwnerMachine(ctx, r.Client, machine.ObjectMeta)
 	if err != nil {
-		return fmt.Errorf("get owner Machine: %s", err)
+		return ctrl.Result{}, fmt.Errorf("get owner Machine: %s", err)
 	}
 	if ownerMachine == nil {
-		return fmt.Errorf("owner Machine is nil")
+		return ctrl.Result{}, fmt.Errorf("owner Machine is nil")
 	}
 
 	ownerCluster, err := capiutil.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
 	if err != nil {
-		return fmt.Errorf("get owner Cluster: %s", err)
+		return ctrl.Result{}, fmt.Errorf("get owner Cluster: %s", err)
 	}
 	if ownerCluster == nil {
-		return fmt.Errorf("owner Cluster is nil")
+		return ctrl.Result{}, fmt.Errorf("owner Cluster is nil")
+	}
+
+	var infrastructureCluster infrastructurev1beta1.VirTinkCluster
+	infrastructureClusterKey := types.NamespacedName{
+		Namespace: ownerCluster.Spec.InfrastructureRef.Namespace,
+		Name:      ownerCluster.Spec.InfrastructureRef.Name,
+	}
+	if err := r.Get(ctx, infrastructureClusterKey, &infrastructureCluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("get infrastructure Cluster: %s", err)
+	}
+	useExternalCluster := infrastructureCluster.Spec.ExternalClusterSecretRef != nil
+	externalCluster, err := externalcluster.NewFromKubeconfigSecretRef(ctx, r.Client, infrastructureCluster.Spec.ExternalClusterSecretRef)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("create external Cluster client from kubeconfig Secret reference: %s", err)
+	}
+	if !machine.DeletionTimestamp.IsZero() {
+		if err := r.reconcileDelete(ctx, machine, externalCluster, useExternalCluster, ownerMachine.Spec.Bootstrap.DataSecretName); err != nil {
+			return ctrl.Result{}, fmt.Errorf("reconcile delete: %s", err)
+		}
+		return ctrl.Result{}, nil
 	}
+
 	if !ownerCluster.Status.InfrastructureReady {
-		return fmt.Errorf("owner Cluster is not ready")
+		return ctrl.Result{}, fmt.Errorf("owner Cluster is not ready")
 	}
 
 	if ownerMachine.Spec.Bootstrap.DataSecretName == nil {
-		return fmt.Errorf("bootstrap data is nil")
+		return ctrl.Result{}, fmt.Errorf("bootstrap data is nil")
+	}
+	if useExternalCluster {
+		var secret corev1.Secret
+		secretKey := types.NamespacedName{
+			Namespace: machine.Namespace,
+			Name:      *ownerMachine.Spec.Bootstrap.DataSecretName,
+		}
+		secretNotFound := false
+		if err := externalCluster.Get(ctx, secretKey, &secret); err != nil {
+			if apierrors.IsNotFound(err) {
+				secretNotFound = true
+			} else {
+				return ctrl.Result{}, fmt.Errorf("get bootstrap Secret from external Cluster: %s", err)
+			}
+		}
+		if secretNotFound {
+			if err := r.Get(ctx, secretKey, &secret); err != nil {
+				return ctrl.Result{}, fmt.Errorf("get bootstrap Secret: %s", err)
+			}
+			secret.ObjectMeta.ResourceVersion = ""
+			secret.ObjectMeta.OwnerReferences = nil
+			if err := externalCluster.Create(ctx, &secret); err != nil {
```

Comment:
VirTink supports embed cloud-init data in VM object, maybe we could consider this option here